src/Platforms/CUDA/CUDAallocator.hpp

```diff
@@ -152,6 +153,53 @@ struct CUDAAllocator
   template<class U>
   static void destroy(U* p)
   {}
+
+  void copyToDevice(T* device_ptr, T* host_ptr, size_t n)
+  {
+    if (hstream_)
+    {
+      cudaErrorCheck(cudaMemcpyAsync(device_ptr, host_ptr, sizeof(T) * n, cudaMemcpyHostToDevice, hstream_),
+                     "cudaMemcpyAsync failed in copyToDevice");
+    }
+    else
+    {
+      cudaErrorCheck(cudaMemcpy(device_ptr, host_ptr, sizeof(T) * n, cudaMemcpyHostToDevice),
+                     "cudaMemcpy failed in copToDevice");
+    }
+  }
+
+  void copyFromDevice(T* host_ptr, T* device_ptr, size_t n)
+  {
+    if (hstream_)
+    {
+      cudaErrorCheck(cudaMemcpyAsync(host_ptr, device_ptr, sizeof(T) * n, cudaMemcpyDeviceToHost, hstream_),
+                     "cudaMemcpyAsync failed in copyFromDevice");
+    }
+    else
+    {
+      cudaErrorCheck(cudaMemcpy(host_ptr, device_ptr, sizeof(T) * n, cudaMemcpyDeviceToHost),
+                     "cudaMemcpy failed in copyFromDevice");
+    }
+  }
+
+  void copyDeviceToDevice(T* to_ptr, size_t n, T* from_ptr)
+  {
+    if (hstream_)
+    {
+      cudaErrorCheck(cudaMemcpyAsync(to_ptr, from_ptr, sizeof(T) * n, cudaMemcpyDeviceToDevice, hstream_),
+                     "cudaMemcpyAsync failed in copyDeviceToDevice");
+    }
+    else
+    {
+      cudaErrorCheck(cudaMemcpy(to_ptr, from_ptr, sizeof(T) * n, cudaMemcpyDeviceToDevice),
+                     "cudaMemcpy failed in copyDeviceToDevice");
+    }
+  }
+  
+  void set_stream(cudaStream_t stream) { hstream_ = stream; }
+
+private:
+  cudaStream_t hstream_ = 0;
```

Comment:
All these copy APIs should be used for testing or performance irrelevant locations. Adding "hstream_" is not a sane design. Prefer remove.

Reply:
I think its much more important to be able to write good test of everything. The API's need to be consistent but openmp target is still 1st class and native c++ second class. Stream is secret state in the openmp runtime so it can't be in the API for transfers. But if any comparison is going to be made between openmp target code and c++ with minimal native vendor code cuda transfers need to be able to set stream.  Perhaps the transfers should be handled by a separate functor, but the allocator is where the device_ptr state is and device_ptr's will definitely have a state relation with streams in future cuda code (there are now stream local allocator calls). And the the openmp runtime should take advantage of this.

Reply:
The transfer calls updateTo/From in OMPallocator is synchronous. OpenMP has nothing to do with streams. In CUDA code if stream is used, i would see it being passed at use not via CUDAallocator.

Reply:
how do we get an asynchronous transfer from openmp target?